portability: Allow static allocation of RTOS objects for CMSIS-RTOSv2 API port

utsavm9 opened the pull request:
Implements [#75626](https://github.com/zephyrproject-rtos/zephyr/issues/75626).

## Overview of changes
* Make the control block types of CMSIS-RTOSv2 object available in a public header for users.
* Handle when the control blocks are passed via `.cb_mem` in all CMSIS-RTOSv2 `os***New()` APIs.
* The files touched did not follow the linter format before the changes, `clang-format` them all.
* Add tests.

## Test Plan
* Ran locally with `west build tests/subsys/portability/cmsis_rtos_v2 --board=nrf52dk/nrf52832`, existing and new tests passed.

## Potential Follow-ups
* If an API like `k_mem_slab_is_good_ptr(struct k_mem_slab *slab, void *mem)` is available, there will not be any need to add `.is_cb_dynamic_allocation` to store this information if a pointer is pointing to an item allocated in a memory slab.
* `.name` can be a `const char *` instead of `char[16]` to save on unnecessarily allocated memory in control blocks.
* Remove max thread limitation in CMSIS-RTOSv2 threads port when using static allocation. RTX5 is the reference implementation of CMSIS-RTOSv2 from ARM, and it does not have this.

utsavm9 committed:
portability: cmsis: CMSIS-RTOSv2 thread port to track dynamic cb

No functionality change, in preparation for allowing
threads with user provided stack and control block.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
portability: cmsis: Support static CMSIS-RTOSv2 control blocks

Do not use memory slabs for the control blocks when
the application provides the memory for it. This
implements manual user-defined allocation memory
management support in CMSIS-RTOSv2 API.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
tests: portability: cmsis: Add tests for CMSIS-RTOSv2 static cb

Every CMSIS-RTOSv2 now supports static allocation
control blocks. Add tests for this.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
tests: portability: cmsis: Run clang-format on CMSIS-RTOSv2 tests

clang-format -i tests/subsys/portability/cmsis_rtos_v2/src/*.c

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 commented:
@nashif Let me know if PR needs to be split or any other changes you want to be made. I didn't really get any comments on the issue either, so I am getting a bit hesitant if improvements to CMSIS-RTOSv2 port can be merged into Zephyr or not.

cfriedt commented:
It would be a bit easier to review if a clang-format, move, and rename, and were split into separate commits (in that order) from the first commit.

I question the need to rename cv2 prefixed things to cmsis_rtos prefixed things. It's doubtful that it's interfering with anything else with the existing prefix and renaming may add an unnecessary cognitive load for reviewers. Maybe it would be best to leave to a follow-up PR.

Otherwise, nothing particularly wrong jumps out at me here, except I am not entirely sure if the follow-up directions are entirely correct.

cfriedt commented:
Maybe prefix commit subject lines with "portability: cmsis: " to be more specific. It's best to follow what was  done in previous commits.

cfriedt commented:
Nothing to do with this PR specifically, but I wonder if @peter-mitsis or maybe @ceolin might comment on userspace-ability.

Perhaps it is not the intent of CMSIS to support userspace, but I guess that CMSIS is only supported in kernel mode today in Zephyr.

From my understanding, userspace should refer to kernel objects via pointer only, outside of static declaration via existing macros. Either that, or there would need to be `__syscall` wrappers to modify control block kernel object data.

Does embedding kernel objects inside of other API structures prevent that API from supporting userspace? For example, do k_mutex objects need to be placed in the same contiguous section to support userspace? In kernel space, there is much more flexibility to initialize and use kernel objects from arbitrary memory.

cfriedt commented:
> .name can be a const char * instead of char[16] to save on unnecessarily allocated memory in control blocks.

The kernel does support thread names, so there shouldn't be any need to duplicate that for CMSIS thread objects. There may be some utility in adding a name field to some other objects, for the purposes of inter-thread communication.

> Remove max thread limitation in CMSIS-RTOSv2 threads port when using static allocation. RTX5 is the reference implementation of CMSIS-RTOSv2 from ARM, and it does not have this.

Zephyr might have additional use cases compared to RTX5. E.g. a mixture of static and dynamic allocation, userspace support, etc.

utsavm9 committed:
portability: cmsis: Run clang-format on CMSIS-RTOSv2

clang-format -i subsys/portability/cmsis_rtos_v2/*.h
clang-format -i subsys/portability/cmsis_rtos_v2/*.c

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
portability: cmsis: Move cmsis wrapper types to public header

This enables the cmsis wrapper types to be declared
statically and then passed along to CMSIS-RTOSv2 APIs,
enabling static allocation of RTOS control blocks
in the subsequent commits.

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 committed:
portability: cmsis: Rename cmsis wrapper types

Namespace this types with `cmsis_rtos` instead of `cv2`

Signed-off-by: Utsav Munendra <utsavm@meta.com>

utsavm9 commented:
@cfriedt Responded to feedback:
* Split the first commit into 3.
* I'll pushback on the need to rename. Once the header is public, user applications will start declaring the wrapper types and then a remane in future will cause breaking changes, so it is better to rename right now. `cv2` makes me think computer vision before CMSIS-RTOSv2, so let us just aptly namespace these types with `cmsis_rtos_`.
* Prefixed commit subject lines with "portability: cmsis:"
* Thanks for the comments on the followups, I will keep them in mind while making the PRs for those follow-ups.

peter-mitsis reviewed (approved):
I'm no CMSIS expert, but this seems reasonable.
